---
title: "astRal"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{astRal}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(astRal)
library(cocoframer)
library(purrr)
library(viridisLite) # optional - nice color palettes
library(dplyr)
library(reshape2)
#library(Seurat)
library(rgl)
library(nat)
library(Matrix)
#library(monocle3)
#library(sf)
#library(NNLM)
library(tricycle)
```

```{r}
#Sagittal Energy Matrix
ccf_v3_coords_file<-system.file("extdata","p56coord_dev_v3_left.RData",package="astRal")
struct_file<-system.file("extdata","p56_dev_structureID_left.RData",package="astRal")
group_file<-system.file("extdata","group.RData",package="astRal")
energy_file<-system.file("extdata","energy_dev_v3_sagittal.RData",package="astRal")
load(ccf_v3_coords_file)
load(struct_file)
load(group_file)
load(energy_file)

```

```{r}
# Get structure ontology annotations
#ga <- get_ccf_grid_annotation()

ontology <- get_mba_ontology()
ontology_df <- flatten_mba_ontology(ontology)

full_ccf_annot<-ccf_annotation(age="Adult",
                               make_array=FALSE,
                               array_spacing=200,
                               coco_reflect=TRUE)

# Recursively identifies ALL ontology structure IDs associated with each voxel
tree_path<-function(id,ontology_df){
  if(id==0){
    return(-1)
  }else{
  path<-c(id)
  while(id!=997){
    id<-ontology_df$parent_structure_id[ontology_df$id==id]
    path<-c(path,id)
  }
  return(path)
  }
}

#energy_mat_dims<-aba_array_dims(resolution=200)

# Add column names to energy matrix
colnames(energy.mat)<-rownames(full_ccf_annot)

# Reduce full annotation to only that subset with signal in Allen in situ database
energy.annot<-energy.mat[,coordID_annotation_dev_v3_left]
reduced.annot<-full_ccf_annot[coordID_annotation_dev_v3_left,]

# build a 3d array of ontology structure acronyms - easier to deal with than IDs
#oa <- array(ontology_df$acronym[match(ga, ontology_df$id)], dim = dim(ga))
```

#Create SpatialExperiment object from Allen in situ dataset
```{r}
se <- NewSpatialExperiment(expression_data=energy.annot,
                           spatial_coords=reduced.annot,
                              barcode_metadata = NULL,
                              gene_metadata = NULL)

assay(se,'logcounts')<-assay(se,'counts')+1

```

## Some sample images of gene expression...
```{r}

spatial_plot_3d(se,color_by="Satb2")
par3d(windowRect = c(20, 30, 400, 400))
movie3d(spin3d(axis = c(1, 0, 0), rpm = 12),fps=20, duration = 5,movie = "Satb2",dir = "/Users/loyalgoff/Downloads/", convert= NULL)

spatial_plot_3d(se,color_by="Otx2")
par3d(windowRect = c(20, 30, 400, 400))
movie3d(spin3d(axis = c(1, 0, 0), rpm = 12),fps=20, duration = 5,movie = "Otx2",dir = "/Users/loyalgoff/Downloads/", convert= NULL)

spatial_plot_3d(se,color_by="Rorb")
par3d(windowRect = c(20, 30, 400, 400))
movie3d(spin3d(axis = c(1, 0, 0), rpm = 12),fps=20, duration = 5,movie = "Rorb",dir = "/Users/loyalgoff/Downloads/", convert= NULL)

spatial_plot_3d(se,color_by="Drd1")
par3d(windowRect = c(20, 30, 400, 400))
movie3d(spin3d(axis = c(1, 0, 0), rpm = 12),fps=20, duration = 5,movie = "Drd1",dir = "/Users/loyalgoff/Downloads/", convert= NULL)

spatial_plot_3d(se,color_by="Calb1")
par3d(windowRect = c(20, 30, 400, 400))
movie3d(spin3d(axis = c(1, 0, 0), rpm = 12),fps=20, duration = 5,movie = "Calb1",dir = "/Users/loyalgoff/Downloads/", convert= NULL)
```

# Tricycle projection
```{r}
se.tricycle<-project_cycle_space(se,exprs_values='logcounts',gname.type='SYMBOL')
se.tricycle<-estimate_cycle_position(se.tricycle)

plot_emb_circle_scale(se.tricycle)
circle_scale_legend()

xlim<-c(1,max(sf::st_bbox(se.tricycle@spatialMap)))
  ylim<-xlim
  zlim<-ylim

hue.colors = c(
        "#2E22EA", "#9E3DFB", "#F86BE2", "#FCCE7B", "#C4E416", "#4BBA0F", "#447D87", "#2C24E9"
    )
  
open3d()
par3d(windowRect = c(20, 30, 1200, 1200))
plot3d(coordinates(se.tricycle),col=myColorRampTricycle(palette=hue.colors,values=colData(se.tricycle)$tricyclePosition),size=10,box=T,axes=T,xlab="",ylab="",zlab="",xlim=xlim,ylim=ylim,zlim=zlim,alpha=0.1)
movie3d(spin3d(axis = c(1, 1, 1), rpm = 6),duration = 10,movie = "Tricycle_Allen_InSitu_P56_Brain",dir = "/Users/loyalgoff/Downloads/", convert= NULL)


structures <- c("root", "HY")
mesh_list <- map(structures, ccf_2017_mesh)

names(mesh_list) <- structures

plot_ccf_meshes(mesh_list,
                fg_structure = c("Hy"),
                fg_alpha = 0.2,
                bg_structure = "root",
                bg_alpha = 0.1,
                style="matte"
                )
par3d(windowRect = c(20, 30, 1200, 1200))
points3d(sf::st_coordinates(se.tricycle@spatialMap),col=myColorRampTricycle(values=colData(se.tricycle)$tricyclePosition),size=10,box=T,axes=T,xlab="",ylab="",zlab="",alpha=0.2, main="Thalamus")
movie3d(spin3d(axis = c(1, 1, 1), rpm = 6),duration = 10,movie = "Tricycle_Allen_InSitu_P56_Brain_Hypothalamus",dir = "/Users/loyalgoff/Downloads/", convert= NULL)
```


# NNMF Patterns
```{r,eval=FALSE}
## NNMF100
#energy.lognorm<-LogNormalize(exprs(se))%>%as.matrix()

# Import existing NNMF120 patterns
#patterns_file <- system.file("extdata","NNMF120.1.RData",package="astRal")
#load("inst/extdata/NNMF120.1.RData")
load("inst/extdata/Energy_imputed_lognorm_CoGAPS120.RData")
#pWeights<-NNMF120.1$H
pWeights<-t(Energy_imputed_lognorm_CoGAPS120@sampleFactors)
colnames(pWeights)<-rownames(pData(se))
rownames(pWeights)<-paste("pattern",c(1:120),sep="_")
patterns.se<-NewSpatialExperiment(expression_data=pWeights,
                           spatial_coords=reduced.annot,
                              barcode_metadata = NULL,
                              gene_metadata = NULL)
```

```{r,eval=FALSE}
open3d()

par3d(windowRect = c(20, 30, 1200, 1200))

n<-20
start<-81

mat <- matrix(1:n,nrow=4)

layout3d(mat, sharedMouse = TRUE)

xlim<-c(1,max(sf::st_bbox(patterns.se@spatialMap)))
  ylim<-xlim
  zlim<-ylim

for (i in c(start:(start+n))) {

    next3d()

    plot3d(sf::st_coordinates(patterns.se@spatialMap),col=myColorRamp(palette=inferno(100),exprs(patterns.se)[i,]),size=3,box=F,axes=F,xlab="",ylab="",zlab="",xlim=xlim,ylim=ylim,zlim=zlim, main=i, alpha=(exprs(patterns.se)[i,]-min(exprs(patterns.se)[i,]))/(max(exprs(patterns.se)[i,])-min(exprs(patterns.se)[i,])))
}
  
rgl.snapshot( "~/Downloads/patterns.png", fmt = "png", top = TRUE )

```

```{r movie, eval=FALSE}
i<-88    #67,70,77,101,88,105,106
open3d()
par3d(windowRect = c(20, 30, 1200, 1200))
plot3d(coordinates(patterns.se),col=myColorRamp(palette=inferno(100),exprs(patterns.se)[i,]),size=10,box=T,axes=T,xlab="",ylab="",zlab="",xlim=xlim,ylim=ylim,zlim=zlim, alpha=(exprs(patterns.se)[i,]-min(exprs(patterns.se)[i,]))/(max(exprs(patterns.se)[i,])-min(exprs(patterns.se)[i,])))
movie3d(spin3d(axis = c(1, 1, 1), rpm = 6),duration = 10,movie = paste0("NNMF120",".1","Pattern_", as.character(i)),dir = "/Users/loyalgoff/Downloads/", convert= NULL)
while (rgl.cur() > 0) {
            rgl.close()
      }
```

```{r merge, eval=FALSE}
open3d()
par3d(windowRect = c(20, 30, 800, 800))
i<-63 # Look at 40(root only),
structures <- c("root", "MY-sen")
mesh_list <- map(structures, ccf_2017_mesh)

names(mesh_list) <- structures

xlim<-c(1,max(sf::st_bbox(patterns.se@spatialMap)))
ylim<-xlim
zlim<-ylim

plot_ccf_meshes(mesh_list,
                fg_structure = c("MY-sen"),
                fg_alpha = 0.2,
                bg_structure = "root",
                style="matte"
                )

points3d(sf::st_coordinates(patterns.se@spatialMap),col=myColorRamp(palette=inferno(100),exprs(patterns.se)[i,]),size=10,box=T,axes=T,xlab="",ylab="",zlab="", alpha=(exprs(patterns.se)[i,]-min(exprs(patterns.se)[i,]))/(max(exprs(patterns.se)[i,])-min(exprs(patterns.se)[i,])))

movie3d(spin3d(axis = c(1, 1, 1), rpm = 12),fps=20, duration = 5,movie = paste0("NNMF120",".1","Pattern_", as.character(i),"_MY-sen"),dir = "/Users/loyalgoff/Downloads/", convert= NULL)
```

# Import scRNA-Seq data and convert to cell_data_set
```{r}
#target_data_path<-"/Volumes/Data/Users/rotations/weifeng/projection/visual_cortex/GSE71585_RefSeq_counts.csv"
target_data_path<-"/Volumes/Data/Users/rotations/weifeng/projection/hypothalamus/GSE87544_Merged_17samples_14437cells_count\ 2.txt"
target_data<-readr::read_tsv(target_data_path)
target_genes<-as.data.frame(target_data[,1])
rownames(target_genes)<-target_genes$Gene

#target_data<-as(as.matrix(target_data[,-1]),"sparseMatrix")
target_data<-target_data[,-1]
target_data<-as.matrix(target_data)

rownames(target_data)<-rownames(target_genes)

cell_info<-data.frame(row.names=colnames(target_data),cell_id=colnames(target_data))

target_data_subset<-target_data[,c(1:100)]

#cds<-new_cell_data_set(target_data,
#                       gene_metadata=target_genes,
#                       cell_metadata=cell_info)

```

# Projection
```{r}
#A.weights<-NNMF120.1$W
A.weights<-Energy_imputed_lognorm_CoGAPS120@featureLoadings
length(intersect(rownames(A.weights),rownames(target_data_subset)))

library(projectR)

cell.projections<-projectR(log10(target_data+1),loadings = A.weights)

# patterns.scaled<-apply(as.matrix(exprs(patterns.se)),1,function(x){
#   scale(x)
# })
# patterns.scaled<-t(patterns.scaled)

patterns.scaled<-as.matrix(t(scale(t(exprs(patterns.se)),center=FALSE)))

#projection.cov <- cov(cell.projections,as.matrix(exprs(patterns.se)))
#projection.dot <- t(cell.projections)%*%as.matrix(exprs(patterns.se))
#projection.dot.scaled <- t(cell.projections)%*%patterns.scaled
projection.cor <- cor(cell.projections,patterns.scaled)


#voxel_idx<-apply(projection.cov,1,which.max)
#voxel_idx<-apply(projection.dot,1,which.max)
voxel_idx<-apply(projection.cor,1,which.max)

open3d()
par3d(windowRect = c(20, 30, 800, 800))
i<-65 # Look at 69 next and then 75, then 95+CTX
structures <- c("root","HIP")
mesh_list <- map(structures, ccf_2017_mesh)

names(mesh_list) <- structures

xlim<-c(1,max(sf::st_bbox(patterns.se@spatialMap)))
ylim<-xlim
zlim<-ylim

plot_ccf_meshes(mesh_list,
                fg_structure = c("HIP"),
                fg_alpha = 0.2,
                bg_structure = "root",
                style="matte"
                )


points3d(sf::st_coordinates(se@spatialMap[as.numeric(names(table(voxel_idx))),]),col=myColorRamp(palette=inferno(100),log10(table(voxel_idx))),size=10,box=F,axes=F)

movie3d(spin3d(axis = c(0, 1, 0), rpm = 12),fps=20, duration = 5,movie = "Imputed_CoGAPS_consensus_patterns_projection_attempt_heatmap_hippocampus",dir = "/Users/loyalgoff/Downloads/", convert= NULL)
```

# One hot encoding of structural features vs pattern usage heatmap
```{r}
full_tree_paths<-lapply(full_ccf_annot$Annot,function(x){tree_path(x,ontology_df)})
full_tree_paths_subset<-full_tree_paths[coordID_annotation_dev_v3_left]
sample(full_tree_paths_subset,5)

# Make 1-hot encoding matrix by pixels
annot.mat<-c()
for(i in unique(ontology_df$id)){
  #print(i)
  idx<-as.numeric(unlist(lapply(full_tree_paths_subset,function(x){i %in% x})))
  annot.mat<-cbind(annot.mat,idx)
}
colnames(annot.mat)<-unique(ontology_df$id)
annot.mat<-as.matrix(annot.mat)

patterns.matrix<-t(scale(t(exprs(patterns.se)),center=FALSE))

pattern.annot.cor<-cor(t(patterns.matrix),annot.mat)

# Create vector of major structures for heatmap
structure_col<-rep("NA",length=dim(pattern.annot.cor)[2])
top_level_ids<-ontology_df[ontology_df$parent_structure_id %in% c(567,343,512),"id"]
for(i in c(1:length(unique(ontology_df$id)))){
  for(t in top_level_ids){
    if(t %in% tree_path(unique(ontology_df$id)[i],ontology_df)){
      name<-ontology_df$name[ontology_df$id==t]
      structure_col[i]<-as.character(name)
    }
  }
}

pos_struct_idx<-apply(pattern.annot.cor,2,function(x){!any(is.na(x))})


cor_cutoff<-0.3
high_cor_val<-apply(pattern.annot.cor,2,function(x){any(x>=cor_cutoff)})
table(high_cor_val)

annotation_col<-data.frame("structure"=structure_col)
rownames(annotation_col)<-unique(ontology_df$id)



library(pheatmap)
plotMat<-pattern.annot.cor[,pos_struct_idx&high_cor_val]
plotAnnotCol<-as.data.frame(annotation_col[pos_struct_idx&high_cor_val,])
colnames(plotAnnotCol)<-"structure"
plotAnnotCol$structure<-as.character(plotAnnotCol$structure)
rownames(plotAnnotCol)<-as.character(unique(ontology_df$id)[pos_struct_idx&high_cor_val])
#rownames(plotAnnotCol)<-unlist(lapply(rownames(plotAnnotCol),function(x){ontology_df$name[ontology_df$id==x]}))
pheatmap(plotMat,annotation_col=plotAnnotCol,cutree_rows=8)

pdf("~/Downloads/structure_by_pattern_heatmap.pdf",width=30,height=20)
  pheatmap(plotMat,annotation_col=plotAnnotCol,cutree_rows=8,border_color="black",border=TRUE)
dev.off()
```

# Plot pattern:structure combo
```{r plot_pattern_and_structure, eval=FALSE}

pat_struct_plot<-function(structure_id,pattern){
  
  open3d()
  par3d(windowRect = c(20, 30, 800, 800))
  i<-pattern # Look at 40(root only),
  print(ontology_df$name[ontology_df$id==structure_id])
  structures <- c("root", ontology_df$acronym[ontology_df$id==structure_id])
  mesh_list <- map(structures, ccf_2017_mesh)
  
  names(mesh_list) <- structures
  
  xlim<-c(1,max(sf::st_bbox(patterns.se@spatialMap)))
  ylim<-xlim
  zlim<-ylim
  
  plot_ccf_meshes(mesh_list,
                  fg_structure = c(ontology_df$acronym[ontology_df$id==structure_id]),
                  fg_alpha = 0.2,
                  bg_structure = "root",
                  style="matte"
                  )
  
  points3d(sf::st_coordinates(patterns.se@spatialMap),col=myColorRamp(palette=inferno(100),exprs(patterns.se)[i,]),size=10,box=T,axes=T,xlab="",ylab="",zlab="", alpha=(exprs(patterns.se)[i,]-min(exprs(patterns.se)[i,]))/(max(exprs(patterns.se)[i,])-min(exprs(patterns.se)[i,])),title=ontology_df$name[ontology_df$id==structure_id])

}

pat_struct_plot(961,68)
pat_struct_plot(754,84)
pat_struct_plot(159,31)
pat_struct_plot(549,85)
pat_struct_plot(856,43)
pat_struct_plot(1097,75)
pat_struct_plot(322,80)
pat_struct_plot(669,88)
pat_struct_plot(726,46)
pat_struct_plot(4,79)
pat_struct_plot(645,25)
pat_struct_plot(909,40)
pat_struct_plot(322,90)
pat_struct_plot(1041,92)

movie3d(spin3d(axis = c(1, 1, 1), rpm = 12),fps=20, duration = 5,movie = "Pattern_85_thalamus",dir = "/Users/loyalgoff/Downloads/", convert= NULL)
```


```




